rust_src/src/lists.rs

```diff
@@ -313,6 +323,58 @@ impl LispCons {
             CHECK_IMPURE(self.0, self._extract() as *mut c_void);
         }
     }
+
+    pub fn equal(
+        self,
+        other: LispCons,
+        kind: equal_kind::Type,
+        depth: i32,
+        ht: LispObject,
+    ) -> bool {
```

Comment:
I dislike this code duplication. However, the iterators are of different type which prevents the logic from being collapsed. The type of iterator and the check for equality for each item is the only difference between the two stanzas.

One option is a macro. But it felt wrong.

It would look like this:

```
if kind == equal_kind::EQUAL_NO_QUIT {
            let mut it1 = LispObject::from(self).iter_tails_unchecked();
            let mut it2 = LispObject::from(other).iter_tails_unchecked();

             cons_equal!(it1, it2,
                         0, Qnil,   // these two are for the check on each item
                         depth + 1, ht) // and these are for when none matched.
} else {
            let mut it1 = LispObject::from(self).iter_tails_noendchecked();
            let mut it2 = LispObject::from(other).iter_tails_noendchecked();

             cons_equal!(it1, it2,
                         depth + 1, ht,   // these two are for the check on each item
                         depth + 1, ht) // and these are for when none matched.
}
```

Reply:
To resolve this, the trait `LispConsIterator` has been added which has `Iterator` has a supertrait.
Also, while cleaning up the Iterators I added a macro to create the Car based Iterators.

The diff is messy. Avoid it and look at the actual Rust code.

Reply:
Nevermind, code is happy now.